pkg/envinfo/url.go

```diff
@@ -0,0 +1,275 @@
+package envinfo
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	devfilev1 "github.com/devfile/api/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/library/pkg/devfile/generator"
+	"github.com/devfile/library/pkg/devfile/parser"
+	"github.com/openshift/odo/pkg/localConfigProvider"
+	"github.com/openshift/odo/pkg/odo/util/validation"
+	"github.com/openshift/odo/pkg/util"
+	"github.com/pkg/errors"
+)
+
+// GetPorts returns the ports, returns default if nil
+func (ei *EnvInfo) GetPorts() []string {
+
+	containerComponents := ei.devfileObj.Data.GetDevfileContainerComponents()
+	portMap := make(map[string]bool)
+
+	var portList []string
+	for _, component := range containerComponents {
+		for _, endpoint := range component.Container.Endpoints {
+			portMap[strconv.FormatInt(int64(endpoint.TargetPort), 10)] = true
+		}
+	}
+
+	for port := range portMap {
+		portList = append(portList, port)
+	}
+
+	sort.Strings(portList)
+	return portList
+}
+
+// CompleteURL completes the given URL with default values
+func (ei *EnvInfo) CompleteURL(url *localConfigProvider.LocalURL) error {
+	if len(url.Path) > 0 && (strings.HasPrefix(url.Path, "/") || strings.HasPrefix(url.Path, "\\")) {
+		if len(url.Path) <= 1 {
+			url.Path = ""
+		} else {
+			// remove the leading / or \ from provided path
+			url.Path = string([]rune(url.Path)[1:])
+		}
+	}
+	// add leading / to path, if the path provided is empty, it will be set to / which is the default value of path
+	url.Path = "/" + url.Path
+
+	containerComponents := ei.devfileObj.Data.GetDevfileContainerComponents()
+	if len(containerComponents) == 0 {
+		return nil
+	}
+
+	if len(url.Container) > 0 {
+		return nil
+	}
+
+	containerPortMap := make(map[int]string)
+	portMap := make(map[string]bool)
+	for _, component := range containerComponents {
+		for _, endpoint := range component.Container.Endpoints {
+			containerPortMap[endpoint.TargetPort] = component.Name
+			portMap[strconv.FormatInt(int64(endpoint.TargetPort), 10)] = true
+		}
+	}
+	if containerName, exist := containerPortMap[url.Port]; exist {
+		if len(url.Container) == 0 {
+			url.Container = containerName
+		}
+	}
+
+	// container is not provided, or the specified port is not being used under any containers
+	// pick the first container to store the new endpoint
+	if len(url.Container) == 0 {
+		url.Container = containerComponents[0].Name
+	}
+
+	if url.Port == -1 {
+		var err error
+		url.Port, err = util.GetValidPortNumber(ei.GetName(), url.Port, ei.GetPorts())
+		if err != nil {
+			return err
+		}
+	}
+
+	// get the name
+	if len(url.Name) == 0 {
+		url.Name = util.GetURLName(ei.GetName(), url.Port)
+	}
+
+	return nil
+}
+
+// ValidateURL validates the given URL
+func (ei *EnvInfo) ValidateURL(url localConfigProvider.LocalURL) error {
+
+	containers, err := generator.GetContainers(ei.devfileObj)
+	if err != nil {
+		return err
+	}
+	if len(containers) == 0 {
```

Comment:
I'm assuming that this condition would be satisfied when there's no `container` components in the devfile. But if that was the case, I think we would have hit some failures in the `Complete` function of url create because it already does things which would not be possible if there were no container components, no?

Reply:
I kept the check in Validate() too in case the function is called without `Complete()` when completion is not required.

Reply:
OK. I think this could be a good candidate for what @girishramnani mentioned in https://github.com/openshift/odo/issues/4313#issuecomment-752042681.

Reply:
Since currently the pattern we follow is complete/validate/run, we should keep them separate for now.

Reply:
Indeed. I was just mentioning it for a reference in the other issue andf or Girish.